### PR TITLE
Integration infrastructure fixes/enhancements

### PR DESCRIPTION
**Pull request by bgeltz**
- Backup cloned git repos and archives automatically
- Allow for args to be passed to the ```git clone``` command; for checking out out specific git branches
- Build optimization: only unpack archives the first time
- And env for SMNG.
- Bug fix for ```gen_slurm.sh```
- Add UNCORE status and ctl registers to default trace_signals for frequency_sweep and monitor
- Add "checkpoint/restart" to ```launch_util.py```
- Add trial retry logic to ```launch_util.py```

### CODE REVIEW

File: integration/apps/build_func.sh [hunk 1]
```diff
@@ -73,9 +74,21 @@ clone_repo_git(){
...
+}
+
+backup_archive() {
+    local ARCHIVE=$1
+    local DIRNAME=$2
+
     # If the user set the cache dir, pack up the source for next time
     if [ -d ${GEOPM_APPS_SOURCES} ]; then
-        tar czvf ${GEOPM_APPS_SOURCES}/${ARCHIVE} ${DIRNAME}
+        if [ ! -f ${ARCHIVE} ]; then
```
Comment (cmcantalupo) [inline integration/apps/build_func.sh:86]: Can you add some comments about why these changes are required, it is not really clear to me what the effect of these changes is.

File: integration/config/smng_env.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+export MPIFORT=mpiifort
+export MPIFC=mpiifort
+export MPIF77=mpiifort
+export MPIF90=mpiifort
+
+export GEOPM_USER_ACCOUNT=h1906
+export GEOPM_SYSTEM_DEFAULT_QUEUE=test
+export GEOPM_SBATCH_EXTRA_LINES="
+#SBATCH --constraint=\"noscratch\"
```
Comment (cmcantalupo) [inline integration/config/smng_env.sh:57]: On mose systems the env.sh file is simply sourced prior to submitting a job, I'm not sure what these #SBATCH lines are doing in that context.

File: integration/experiment/frequency_sweep/frequency_sweep.py [hunk 3]
```diff
@@ -86,7 +86,8 @@ def report_signals():
 
 
 def trace_signals():
-    return []
+    return ['MSR::UNCORE_PERF_STATUS:FREQ@package', 'MSR::UNCORE_RATIO_LIMIT:MAX_RATIO@package',
```
Comment (cmcantalupo) [inline integration/experiment/frequency_sweep/frequency_sweep.py:89]: I don't think traces are enabled for the frequency sweep run by default.

File: integration/experiment/uncore_frequency_sweep/uncore_frequency_sweep.py [hunk 4]
```diff
@@ -95,7 +95,8 @@ def report_signals():
 
 
 def trace_signals():
-    return []
+    return ['MSR::UNCORE_PERF_STATUS:FREQ@package', 'MSR::UNCORE_RATIO_LIMIT:MAX_RATIO@package',
```
Comment (cmcantalupo) [inline integration/experiment/uncore_frequency_sweep/uncore_frequency_sweep.py:98]: Are traces enabled?

File: integration/experiment/launch_util.py [hunk 5]
```diff
@@ -93,17 +94,20 @@ def launch_run(agent_conf, app_conf, run_id, output_dir, extra_cli_args,
     num_ranks = app_conf.get_rank_per_node() * num_nodes
     cpu_per_rank = app_conf.get_cpu_per_rank()
 
-    launcher = geopmpy.launcher.Factory().create(argv,
-                                                 num_node=num_nodes,
-                                                 num_rank=num_ranks,
-                                                 cpu_per_rank=cpu_per_rank)
-    launcher.run()
+    # Attempt to run if there are no valid report files for the current configuration
+    if not (os.path.isfile(report_path) and os.path.getsize(report_path) > 0):
```
Comment (cmcantalupo) [inline integration/experiment/launch_util.py:98]: I thought this was managed by the creation of the machine file.  You cannot create a machine file if reports are present...

**Review by cmcantalupo — COMMENTED** [review]
Some quick feedback.  Might have to sort out some of the script changes over slack.

File: integration/apps/build_func.sh (hunk 1 above)
Comment (bgeltz) [reply integration/apps/build_func.sh:86]: This one is easier if you use the split view instead of unified.  I refactored this code that was responsible for backing up downloads in GEOPM_APPS_SOURCES.  Previously there were 2 different backup paths, one in clone_repo_git() and one in get_archive().

They were doing almost the same operations, so I did an extract refactor so all the backup code is in the same place.

backup_archive() is just a refactor.  No functional change.

File: integration/config/smng_env.sh (hunk 2 above)
Comment (bgeltz) [reply integration/config/smng_env.sh:57]: These lines are used when SBATCH scripts are generated using gen_slurm.sh.  This is the only time they are used.

File: integration/experiment/frequency_sweep/frequency_sweep.py (hunk 3 above)
Comment (bgeltz) [reply integration/experiment/frequency_sweep/frequency_sweep.py:89]: They are not, that's true.  I was hoping to get this in since it's very commonly requested when we start digging into data.  When someone turns traces on, it would be nice to not have to remember to also add the UNCORE fields.

File: integration/experiment/launch_util.py (hunk 5 above)
Comment (bgeltz) [reply integration/experiment/launch_util.py:98]: That has not been my experience with the runs I did in December.  This was required.

File: integration/experiment/uncore_frequency_sweep/uncore_frequency_sweep.py (hunk 4 above)
Comment (bgeltz) [reply integration/experiment/uncore_frequency_sweep/uncore_frequency_sweep.py:98]: No, not by default.  I was hoping to set the trace_signals() up for when they are enabled.

**Review by cmcantalupo — APPROVED** [review]
